modules/meteroid/src/api_rest/auth.rs

```diff
@@ -0,0 +1,208 @@
+use common_grpc::middleware::common::auth::API_KEY_HEADER;
+
+use axum::http::StatusCode;
+use axum::response::Response;
+use cached::proc_macro::cached;
+use common_grpc::middleware::server::auth::api_token_validator::ApiTokenValidator;
+use http::{HeaderMap, Request};
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tower::Service;
+use tower_layer::Layer;
+use tracing::{error, log};
+
+use common_grpc::middleware::common::filters::Filter;
+
+use common_grpc::middleware::server::auth::{AuthenticatedState, AuthorizedAsTenant};
+use meteroid_store::repositories::api_tokens::ApiTokensInterface;
+use meteroid_store::Store;
+use uuid::Uuid;
+
+#[allow(unused)]
+#[derive(Debug)]
+pub struct AuthStatus {
+    status: StatusCode,
+    msg: Option<String>,
+}
+
+#[derive(Clone)]
+pub struct ApiAuthMiddleware<S> {
+    inner: S,
+    filter: Option<Filter>,
+    store: Store,
+}
+
+#[derive(Clone)]
+pub struct ExternalApiAuthLayer {
+    store: Store,
+    filter: Option<Filter>,
+}
+
+impl ExternalApiAuthLayer {
+    #[allow(clippy::new_without_default)]
+    pub fn new(store: Store) -> Self {
+        ExternalApiAuthLayer {
+            store,
+            filter: None,
+        }
+    }
+
+    #[must_use]
+    pub fn filter(mut self, filter: Filter) -> Self {
+        self.filter = Some(filter);
+        self
+    }
+}
+
+impl<S> Layer<S> for ExternalApiAuthLayer {
+    type Service = ApiAuthMiddleware<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        ApiAuthMiddleware {
+            inner,
+            filter: self.filter,
+            store: self.store.clone(),
+        }
+    }
+}
+
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for ApiAuthMiddleware<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>> + Clone + Send + 'static,
+    S::Future: Send + 'static,
+    ReqBody: Default + Send + 'static,
+    ResBody: Default + Send + 'static,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, mut request: Request<ReqBody>) -> Self::Future {
+        if !self.filter.map_or(true, |f| f(request.uri().path())) {
+            return Box::pin(self.inner.call(request));
+        }
+
+        // This is necessary because tonic internally uses `tower::buffer::Buffer`.
+        // See https://github.com/tower-rs/tower/issues/547#issuecomment-767629149
+        // for details on why this is necessary
```

Comment:
we are not using tonic for rest so please either adjust the comment, remove it and change the code accordingly.